File: server/src/routes/reactions.rs
use axum::extract::{Path, Query, State};
use axum::response::{IntoResponse, Response};
use axum::Json;

use crate::auth::AuthUser;
use crate::db;
use crate::error::AppError;
use crate::models::User;
use crate::state::AppState;
use crate::views::room::{ReactionBarFragment, ReactionSurface, ReactionView};
use crate::views::{html, Html};
use crate::ws::events::ChatEvent;

/// LC-595: `?surface=thread` on the reaction endpoints. Absent (the timeline,
/// and every pre-LC-595 caller) deserializes to the default, so existing URLs
/// keep working untouched.
#[derive(Debug, Default, serde::Deserialize)]
pub struct SurfaceQuery {
    #[serde(default)]
    pub surface: ReactionSurface,
}

// percent_encoding is already a transitive dep via the workspace; the picker
// uses it to URL-encode shortcodes/glyphs into the reactions path segment.

/// LC-149: every reaction handler must verify the caller can access the
/// message's room before touching it, exactly like the pinned/bookmark/poll
/// handlers. Without this a caller could react in (or enumerate the custom
/// emoji of) a private room / DM / enclave they are not a member of.
async fn require_room_access(state: &AppState, user: &User, room_id: i64) -> Result<(), AppError> {
    let is_admin = user.role == "admin";
    if !db::chat::is_room_accessible(&state.chat, room_id, &user.id, is_admin).await? {
        return Err(AppError::Forbidden);
    }
    Ok(())
}

/// Upper bound on a reaction token. Legitimate values are a single unicode
/// emoji (a few bytes) or a `:shortcode:` (shortcode is `[a-z0-9_]{2,32}`, so
/// <= 34 bytes). 64 bytes is generous headroom while blocking junk rows from
/// the raw `{emoji}` path segment (LC-149 / audit S3).
const MAX_REACTION_BYTES: usize = 64;

fn is_valid_reaction(emoji: &str) -> bool {
    !emoji.is_empty() && emoji.len() <= MAX_REACTION_BYTES && !emoji.chars().any(|c| c.is_control())
}

/// LC-274: minimal HTML-attribute escaping for values inlined into the raw
/// picker markup (there is no Askama template on this path to auto-escape).
fn attr_escape(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('"', "&quot;")
        .replace('<', "&lt;")
}

/// Number of frequent emoji to return. A handful more than the quick-react
/// bar shows (QCAP=5 in layout.html) so the client can dedupe against recents
/// and still have a full row.
const FREQUENT_LIMIT: i64 = 8;

/// GET /api/reactions/frequent
/// LC-477: the caller's most-used Unicode reaction emoji as a JSON array,
/// most-frequent first. The quick-react bar (LC-302) fetches this once and uses
/// it as the primary seed so the one-tap row reflects the user's real habits
/// (cross-device, frequency-ranked) rather than only device-local MRU. No room
/// scope: the result is the caller's own aggregate across every room they have
/// reacted in, so no per-room access check applies.
pub async fn get_frequent(
    State(state): State<AppState>,
    AuthUser(user): AuthUser,
) -> Result<Json<Vec<String>>, AppError> {
    let emojis = db::chat::top_reaction_emojis(&state.chat, &user.id, FREQUENT_LIMIT).await?;
    Ok(Json(emojis))
}

/// POST /messages/:message_id/reactions/:emoji
/// Toggle the caller's reaction for the given emoji on the given message.
/// Broadcasts a ReactionAdded or ReactionRemoved event to subscribers and
/// returns the updated reaction-bar fragment for the caller's tab.
pub async fn toggle_reaction(
    State(state): State<AppState>,
    AuthUser(user): AuthUser,
    Path((message_id, emoji)): Path<(i64, String)>,
    Query(SurfaceQuery { surface }): Query<SurfaceQuery>,
) -> Result<Html, AppError> {
    let m = db::chat::get_message(&state.chat, message_id)
        .await?
        .ok_or(AppError::NotFound)?;
    require_room_access(&state, &user, m.room_id).await?;
    if !is_valid_reaction(&emoji) {
        return Err(AppError::BadRequest("invalid reaction".into()));
    }

    let added = db::chat::toggle_reaction(&state.chat, message_id, &user.id, &emoji).await?;
    let event = if added {
        ChatEvent::ReactionAdded {
            message_id,
            room_id: m.room_id,
            emoji: emoji.clone(),
            user_id: user.id.clone(),
        }
    } else {
        ChatEvent::ReactionRemoved {
            message_id,
            room_id: m.room_id,
            emoji: emoji.clone(),
            user_id: user.id.clone(),
        }
    };
    state.hub.broadcast_to_room(m.room_id, &event);

    // LC-75: publish reaction.added to outgoing webhooks (additions only).
    // LC-78: actor describes the REACTOR (not the message author). v1 only
    // allows real users to react, so this is always kind="user"; the field
    // is included for symmetry with message.* events and so a future
    // bridge-reactions endpoint can swap it without changing the schema.
    if added {
        crate::outgoing::enqueue(
            &state.chat,
            "reaction.added",
            m.room_id,
            serde_json::json!({
                "message_id": message_id,
                "emoji": emoji,
                "user_id": user.id,
                "actor": super::outgoing_actor(&user.id, None, None, None, None),
            }),
        )
        .await;

        // LC-495: run reaction_added workflow automations off the request path
        // (additions only). A matched rule posts as the automation bot; bot
        // reactors are ignored and the engine never re-enters itself.
        let st = state.clone();
        let reactor = user.clone();
        let emoji_owned = emoji.clone();
        let rid = m.room_id;
        tokio::spawn(async move {
            crate::automations::on_reaction_added(&st, rid, &reactor, &emoji_owned).await;
        });
    }

    let custom_emojis =
        db::custom_emojis::refs_for_room_and_user(&state.chat, m.room_id, &user.id).await?;
    let counts = db::chat::list_reactions(&state.chat, message_id, &user.id).await?;
    let reactor_titles = super::build_reactor_titles(&state, &counts).await;
    let reactions: Vec<ReactionView> = counts
        .into_iter()
        .zip(reactor_titles)
        .map(|(r, title)| {
            ReactionView::new(r.emoji, r.count, r.reacted_by_me, title, &custom_emojis)
        })
        .collect();
    // LC-595: answer in the CALLER's id namespace. A chip clicked in the thread
    // panel must be replaced by markup carrying the thread's id; responding with
    // the timeline's would recreate the duplicate id LC-553 removed.
    let fragment = ReactionBarFragment {
        message_id,
        reactions: &reactions,
        surface,
        oob: false,
    };
    html(&fragment)
}

/// LC-389: build one react button for a Unicode emoji. The markup is byte-for-
/// byte the LC-384 shape (same `data-lc-emoji-name` filter hook, same
/// `hx-post`/`hx-target`/`hx-swap` the LC-288 recorder + close-on-react JS key
/// on), so expanding the set needed no JS change. `name` is the search keyword
/// string (human name + every shortcode), attribute-escaped.
fn push_unicode_button(
    buf: &mut String,
    message_id: i64,
    glyph: &str,
    name: &str,
    surface: ReactionSurface,
) {
    let encoded = percent_encoding::utf8_percent_encode(glyph, percent_encoding::NON_ALPHANUMERIC);
    // LC-595: the picker is opened FROM a surface and reacts back INTO it, so
    // every button carries that surface in both its URL and its target.
    let prefix = surface.id_prefix();
    let suffix = surface.url_suffix();
    buf.push_str(&format!(
        r##"<button type="button" data-lc-emoji-name="{name}" hx-post="/messages/{message_id}/reactions/{encoded}{suffix}" hx-target="#{prefix}reactions-{message_id}" hx-swap="outerHTML" class="lc-emoji-cell">{glyph}</button>"##,
    ));
}

/// GET /messages/:message_id/reactions/picker
/// Return the floating emoji picker. LC-389: the full Unicode set organized
/// into the eight standard categories (browsable via the tab strip + scroll)
/// plus the message's enclave custom emojis. The LC-274 filter searches name +
/// shortcodes across the whole set; the LC-288 recent row sits on top. DMs and
/// non-enclave rooms see the Unicode set without a Custom section.
pub async fn get_picker(
    State(state): State<AppState>,
    AuthUser(user): AuthUser,
    Path(message_id): Path<i64>,
    Query(SurfaceQuery { surface }): Query<SurfaceQuery>,
) -> Result<Response, AppError> {
    // LC-149: gate before disclosing the room's custom-emoji inventory (and
    // before the handler doubles as an unauthenticated message-existence
    // oracle). 404 the message first so a non-member cannot distinguish
    // "no such message" from "no access".
    let m = db::chat::get_message(&state.chat, message_id)
        .await?
        .ok_or(AppError::NotFound)?;
    require_room_access(&state, &user, m.room_id).await?;

    // LC-389: render the full set once per open. ~1900 buttons; reserve so the
    // build is a single allocation rather than a growth cascade.
    let mut tabs = String::new();
    let mut sections = String::with_capacity(400 * 1024);
    for (group, slug, key, tab_glyph) in crate::emoji_catalog::GROUPS {
        let label = attr_escape(&crate::i18n::translate_current(key));
        tabs.push_str(&format!(
            r##"<button type="button" data-lc-emoji-tab="{slug}" aria-label="{label}" title="{label}" class="lc-emoji-tab">{tab_glyph}</button>"##,
        ));
        sections.push_str(&format!(
            r##"<section class="lc-emoji-section" data-lc-emoji-cat="{slug}"><div class="lc-emoji-cat-label">{label}</div><div class="lc-emoji-cat-grid">"##,
        ));
        for emoji in group.emojis() {
            // Keyword string (name + every shortcode) is shared with the
            // composer picker, attribute-escaped against a stray quote.
            let name = attr_escape(&crate::emoji_catalog::keywords(emoji));
            push_unicode_button(&mut sections, message_id, emoji.as_str(), &name, surface);
        }
        sections.push_str("</div></section>");
    }

    // LC-389: per-enclave custom emojis as their own trailing section + tab.
    // Message-specific image buttons, unchanged from LC-384 (the LC-288 recorder
    // deliberately skips `:shortcode:` reactions, so these stay out of Recent).
    let custom =
        db::custom_emojis::refs_for_room_and_user(&state.chat, m.room_id, &user.id).await?;
    if !custom.is_empty() {
        let label = attr_escape(&crate::i18n::translate_current(
            "partials-reaction-cat-custom",
        ));
        tabs.push_str(&format!(
            r##"<button type="button" data-lc-emoji-tab="custom" aria-label="{label}" title="{label}" class="lc-emoji-tab">⭐</button>"##,
        ));
        sections.push_str(&format!(
            r##"<section class="lc-emoji-section" data-lc-emoji-cat="custom"><div class="lc-emoji-cat-label">{label}</div><div class="lc-emoji-cat-grid">"##,
        ));
        for emoji in &custom {
            let key = format!(":{}:", emoji.shortcode);
            let encoded =
                percent_encoding::utf8_percent_encode(&key, percent_encoding::NON_ALPHANUMERIC);
            // shortcode charset is [a-z0-9_]{2,32}, validated on insert,
            // so it is safe to inline into the markup without escaping.
            sections.push_str(&format!(
                r##"<button type="button" data-lc-emoji-name="{code}" hx-post="/messages/{id}/reactions/{enc}{suffix}" hx-target="#{prefix}reactions-{id}" hx-swap="outerHTML" class="lc-emoji-cell" title=":{code}:"><img class="h-5 w-5 inline-block align-text-bottom" src="/api/emojis/{eid}" alt=":{code}:"></button>"##,
                id = message_id,
                enc = encoded,
                prefix = surface.id_prefix(),
                suffix = surface.url_suffix(),
                code = emoji.shortcode,
                eid = emoji.id,
            ));
        }
        sections.push_str("</div></section>");
    }

    // LC-274: localized filter placeholder, attribute-escaped (translations are
    // operator/maintainer-controlled, but escape defensively so a future string
    // with a quote cannot break out of the attribute).
    let filter_label = attr_escape(&crate::i18n::translate_current("partials-reaction-filter"));
    // LC-288: empty recent-emoji row; the layout IIFE fills it from localStorage
    // when the picker opens (hidden when there is no history).
    let recent_label = attr_escape(&crate::i18n::translate_current("partials-reaction-recent"));
    // LC-384: the picker is a floating card rendered into the fixed-position
    // #lc-reaction-popover host (positioned by the layout.html LC-384 script).
    // The id MUST stay `picker-{id}` - the LC-288 recent-row JS derives the
    // message id from it and the LC-274 filter keys on `[id^="picker-"]`. The
    // close `x` dismisses the host client-side (data-lc-reaction-close).
    let close_label = attr_escape(&crate::i18n::translate_current("reminders-close"));
    let body = format!(
        r##"<div id="picker-{message_id}" class="w-72 rounded-lg border border-border bg-surface-elevated shadow-lg overflow-hidden">
  <div class="flex items-center gap-1 p-2 border-b border-border">
    <input data-lc-emoji-filter type="text" autocomplete="off" autofocus aria-label="{filter_label}" placeholder="{filter_label}" class="flex-1 min-w-0 rounded-md border border-border bg-surface-sunken px-2 py-1 text-xs focus:outline-none focus:border-accent focus:ring-1 focus:ring-accent">
    <button type="button" data-lc-reaction-close aria-label="{close_label}" class="shrink-0 inline-flex items-center justify-center h-6 w-6 rounded-md text-content-muted hover:bg-surface-sunken hover:text-content">&times;</button>
  </div>
  <div data-lc-emoji-recent data-lc-recent-label="{recent_label}" class="hidden flex flex-wrap items-center gap-1 px-2 py-1.5 text-xs text-content-muted border-b border-border"></div>
  <div data-lc-emoji-tabs class="lc-emoji-tabs">{tabs}</div>
  <div data-lc-emoji-grid class="lc-emoji-scroll">{sections}</div>
</div>"##,
    );
    Ok(axum::response::Html(body).into_response())
}

/// GET /messages/:message_id/reactions/cancel
/// Replace the picker with the add-reaction button again. LC-377: this mirrors
/// the smiley pill rendered by `partials/reaction_bar.html`. The picker replaced
/// the add button in place (`hx-target="this"`), so after a cancel the user has
/// just been interacting with this message's reactions; the button is rendered
/// visible (not hover-gated) until the next full re-render reapplies the gate.
pub async fn cancel_picker(Path(message_id): Path<i64>) -> Response {
    let add_label = attr_escape(&crate::i18n::translate_current("partials-reaction-add"));
    let body = format!(
        r##"<button hx-get="/messages/{message_id}/reactions/picker" hx-target="this" hx-swap="outerHTML" aria-label="{add_label}" data-lc-tip="{add_label}" data-lc-tip-pos="top" aria-haspopup="dialog" class="lc-react-add inline-flex items-center justify-center h-6 rounded-full border border-border bg-surface-sunken px-1.5 text-content-muted hover:bg-surface-elevated hover:text-content"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" class="h-4 w-4" aria-hidden="true"><path fill-rule="evenodd" d="M10 18a8 8 0 100-16 8 8 0 000 16zM7 9a1 1 0 100-2 1 1 0 000 2zm7-1a1 1 0 11-2 0 1 1 0 012 0zm-.464 5.535a1 1 0 10-1.415-1.414 3 3 0 01-4.242 0 1 1 0 00-1.415 1.414 5 5 0 007.072 0z" clip-rule="evenodd"/></svg></button>"##,
    );
    axum::response::Html(body).into_response()
}

File: server/assets/reactions.js
// LC-553: reaction picker runtime. The reaction bar + get_picker route were
// authored around a shared floating picker host (#lc-reaction-popover) and a
// set of data hooks (tabs / filter / recent / close), but the script that made
// them work was never in the tree - so the picker could not open and reacting
// via the UI did nothing. This restores that behavior: open the picker into the
// fixed host, position it by the trigger, switch category tabs, filter by name,
// keep a recent-emoji MRU, and close on react / Escape / outside-click.
(function () {
  'use strict';
  var HOST_ID = 'lc-reaction-popover';
  var RECENT_KEY = 'lc-react-recent';
  var RECENT_MAX = 16;
  var lastTrigger = null;

  function host() { return document.getElementById(HOST_ID); }
  function isOpen() { var h = host(); return h && !h.classList.contains('hidden'); }
  function close() {
    var h = host();
    if (!h) return;
    h.classList.add('hidden');
    h.replaceChildren();
    if (lastTrigger && lastTrigger.focus) { try { lastTrigger.focus(); } catch (e) {} }
    lastTrigger = null;
  }

  // Remember which [data-lc-reaction-open] opened the picker (capture phase, so
  // it is recorded before htmx fires the hx-get that swaps the host).
  document.addEventListener('click', function (e) {
    var t = e.target.closest && e.target.closest('[data-lc-reaction-open]');
    if (t) lastTrigger = t;
  }, true);

  // Position the fixed host near the trigger: prefer above and right-aligned,
  // flip below when there is not enough room, and clamp into the viewport.
  function position() {
    var h = host();
    if (!h || !lastTrigger) return;
    var card = h.firstElementChild;
    if (!card) return;
    var vw = window.innerWidth, vh = window.innerHeight, m = 8;
    // LC-598 follow-up: cap the picker to the viewport so a tall card opened near
    // a screen edge is never clipped off-screen with no way to reach the hidden
    // rows. Shrink the inner emoji grid (the scrollable part) rather than the whole
    // card, so the filter box and category tabs stay put and only the grid scrolls.
    // Clear the previous clamp first so a resize back to a tall window restores the
    // natural height before re-measuring.
    var grid = card.querySelector('[data-lc-emoji-grid]');
    if (grid) grid.style.maxHeight = '';
    var overflow = card.offsetHeight - (vh - 2 * m);
    if (grid && overflow > 0) {
      grid.style.maxHeight = Math.max(96, grid.offsetHeight - overflow) + 'px';
    }
    var tr = lastTrigger.getBoundingClientRect();
    var cw = card.offsetWidth, ch = card.offsetHeight;
    var top = tr.top - ch - 6;
    if (top < m) top = tr.bottom + 6;
    if (top + ch > vh - m) top = Math.max(m, vh - ch - m);
    var left = tr.right - cw;
    if (left < m) left = m;
    if (left + cw > vw - m) left = Math.max(m, vw - cw - m);
    h.style.top = top + 'px';
    h.style.left = left + 'px';
  }

  // Recent MRU: stored as an array of glyph strings; custom (image) emojis are
  // skipped (their button text is empty) to keep the row Unicode-only.
  function readRecent() {
    try { return JSON.parse(localStorage.getItem(RECENT_KEY) || '[]'); } catch (e) { return []; }
  }
  function recordRecent(btn) {
    var glyph = (btn.textContent || '').trim();
    if (!glyph || glyph.length > 6) return; // skip empty / image buttons
    var arr = readRecent().filter(function (g) { return g !== glyph; });
    arr.unshift(glyph);
    arr = arr.slice(0, RECENT_MAX);
    try { localStorage.setItem(RECENT_KEY, JSON.stringify(arr)); } catch (e) {}
  }
  function renderRecent(card) {
    var row = card.querySelector('[data-lc-emoji-recent]');
    if (!row) return;
    var arr = readRecent();
    if (!arr.length) { row.classList.add('hidden'); return; }
    var cells = card.querySelectorAll('[data-lc-emoji-name]');
    row.replaceChildren();
    var label = document.createElement('span');
    label.textContent = row.getAttribute('data-lc-recent-label') || '';
    label.className = 'mr-1';
    row.appendChild(label);
    var added = 0;
    arr.forEach(function (g) {
      var src = Array.prototype.find.call(cells, function (c) { return c.textContent.trim() === g; });
      if (!src) return;
      row.appendChild(src.cloneNode(true)); // carries hx-post/hx-target
      added++;
    });
    if (!added) { row.classList.add('hidden'); return; }
    if (window.htmx) window.htmx.process(row); // wire the cloned buttons
    row.classList.remove('hidden');
  }

  function initPicker(h) {
    var card = h.querySelector('[id^="picker-"]');
    if (!card) return;
    var sections = card.querySelectorAll('[data-lc-emoji-cat]');
    var tabs = card.querySelectorAll('[data-lc-emoji-tab]');
    var grid = card.querySelector('[data-lc-emoji-grid]');

    function showCat(slug) {
      sections.forEach(function (s) {
        s.style.display = (s.getAttribute('data-lc-emoji-cat') === slug) ? '' : 'none';
      });
      tabs.forEach(function (t) {
        t.classList.toggle('is-active', t.getAttribute('data-lc-emoji-tab') === slug);
      });
      if (grid) grid.scrollTop = 0;
    }
    if (tabs.length) {
      tabs.forEach(function (t) {
        t.addEventListener('click', function () { showCat(t.getAttribute('data-lc-emoji-tab')); });
      });
      showCat(tabs[0].getAttribute('data-lc-emoji-tab'));
    }

    renderRecent(card);

    var filter = card.querySelector('[data-lc-emoji-filter]');
    if (filter) {
      filter.addEventListener('input', function () {
        var q = filter.value.trim().toLowerCase();
        var cells = card.querySelectorAll('[data-lc-emoji-name]');
        if (!q) {
          // back to tab mode
          var active = card.querySelector('[data-lc-emoji-tab].is-active');
          cells.forEach(function (c) { c.style.display = ''; });
          if (tabs.length) showCat((active || tabs[0]).getAttribute('data-lc-emoji-tab'));
          return;
        }
        // filtering spans all categories: reveal every section, hide non-matches
        sections.forEach(function (s) { s.style.display = ''; });
        cells.forEach(function (c) {
          var n = (c.getAttribute('data-lc-emoji-name') || '').toLowerCase();
          c.style.display = n.indexOf(q) >= 0 ? '' : 'none';
        });
      });
    }

    var x = card.querySelector('[data-lc-reaction-close]');
    if (x) x.addEventListener('click', close);

    // React + close. htmx fires the hx-post on this same click; defer the close
    // so the request is already dispatched before the host is emptied.
    card.addEventListener('click', function (e) {
      var b = e.target.closest && e.target.closest('button[hx-post]');
      if (!b) return;
      recordRecent(b);
      setTimeout(close, 0);
    });
  }

  // ── Quick-react bar ───────────────────────────────────────────────────────
  // LC-553: every message row carries a `data-lc-quick-bar` placeholder that was
  // never populated (it shipped `hidden` with no filler), so the one-tap row
  // never appeared. Seed it from the caller's frequency-ranked emoji
  // (/api/reactions/frequent, cross-device) merged with the device-local MRU,
  // topped up with sensible defaults, and fill each row lazily on first hover.
  var QCAP = 5;
  var QUICK_DEFAULTS = ['👍', '🎉', '❤️', '😄', '👀'];
  var frequentSeed = null;
  function seedFrequent() {
    if (frequentSeed) return frequentSeed;
    frequentSeed = fetch('/api/reactions/frequent', { headers: { accept: 'application/json' } })
      .then(function (r) { return r.ok ? r.json() : []; })
      .catch(function () { return []; });
    return frequentSeed;
  }
  function quickList(freq) {
    var out = [];
    function push(g) { if (g && out.indexOf(g) < 0) out.push(g); }
    (Array.isArray(freq) ? freq : []).forEach(push);
    readRecent().forEach(push);
    QUICK_DEFAULTS.forEach(push);
    return out.slice(0, QCAP);
  }
  function fillQuick(bar) {
    if (bar.getAttribute('data-lc-quick-filled') === '1') return;
    var row = bar.closest('[id^="msg-"]');
    if (!row) return;
    var mid = row.id.replace('msg-', '');
    if (!mid) return;
    bar.setAttribute('data-lc-quick-filled', '1');
    seedFrequent().then(function (freq) {
      var list = quickList(freq);
      if (!list.length) return;
      bar.replaceChildren();
      list.forEach(function (g) {
        var b = document.createElement('button');
        b.type = 'button';
        b.className = 'lc-msg-action';
        b.setAttribute('hx-post', '/messages/' + mid + '/reactions/' + encodeURIComponent(g));
        b.setAttribute('hx-target', '#reactions-' + mid);
        b.setAttribute('hx-swap', 'outerHTML');
        b.setAttribute('aria-label', g);
        b.textContent = g;
        bar.appendChild(b);
      });
      if (window.htmx) window.htmx.process(bar);
      bar.classList.remove('hidden');
      bar.classList.add('inline-flex');
    });
  }
  function quickFromEvent(e) {
    var row = e.target.closest && e.target.closest('[id^="msg-"]');
    if (!row) return;
    var q = row.querySelector('[data-lc-quick-bar]');
    if (q) fillQuick(q);
  }
  document.addEventListener('pointerover', quickFromEvent, true);
  document.addEventListener('focusin', quickFromEvent, true);
  // A one-tap quick react counts toward the recent MRU too.
  document.addEventListener('click', function (e) {
    var b = e.target.closest && e.target.closest('[data-lc-quick-bar] button[hx-post]');
    if (b) recordRecent(b);
  }, true);

  // Reveal + position + wire the picker once htmx swaps it into the host.
  document.body.addEventListener('htmx:afterSwap', function (e) {
    if (!e.target || e.target.id !== HOST_ID) return;
    var h = host();
    if (!h || !h.firstElementChild) return;
    h.classList.remove('hidden');
    position();
    initPicker(h);
    var f = h.querySelector('[data-lc-emoji-filter]');
    if (f) f.focus();
  });

  // Close on outside pointerdown (but not on a trigger - htmx reopens it) and Escape.
  document.addEventListener('pointerdown', function (e) {
    if (!isOpen()) return;
    var h = host();
    if (h.contains(e.target)) return;
    if (e.target.closest && e.target.closest('[data-lc-reaction-open]')) return;
    close();
  });
  document.addEventListener('keydown', function (e) {
    if (e.key === 'Escape' && isOpen()) { e.stopPropagation(); close(); }
  });
  // Reposition a still-open picker if the timeline scrolls under it.
  window.addEventListener('scroll', function () { if (isOpen()) position(); }, true);
})();
